File: csharp/core/BaseClient.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Text;

using AlibabaCloud.Apigateway.Utils;

using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

using Tea;

namespace AlibabaCloud.Apigateway
{
    public class BaseClient
    {
        private string _regionId;
        private string _protocol;
        private string _domain;
        private string _appSecret;

        protected string _token;
        protected string _stage;
        protected string _appKey;

        public BaseClient(Dictionary<string, object> config)
        {
            _regionId = DictUtils.GetDicValue(config, "regionId").ToSafeString();
            _protocol = DictUtils.GetDicValue(config, "protocol").ToSafeString();
            _domain = DictUtils.GetDicValue(config, "domain").ToSafeString();
            _token = DictUtils.GetDicValue(config, "token").ToSafeString();
            _appKey = DictUtils.GetDicValue(config, "appKey").ToSafeString();
            _appSecret = DictUtils.GetDicValue(config, "appSecret").ToSafeString();
        }

        protected Dictionary<string, object> _readAsJSON(TeaResponse response)
        {
            string body = TeaCore.GetResponseBody(response);
            Dictionary<string, object> dic = new Dictionary<string, object>();
            JObject jObj = JObject.Parse(body);
            dic = (Dictionary<string, object>) ReadJsonUtil.DeserializeToDic(jObj);
            return dic;
        }

        protected bool _isStatusCode(TeaResponse response, int code)
        {
            return response.StatusCode == code;
        }

        protected string _toJSONString(Dictionary<string, object> dict)
        {
            return JsonConvert.SerializeObject(dict);
        }

        protected string _getUUID()
        {
            return Guid.NewGuid().ToString();
        }

        protected string _getTimestamp()
        {
            return DateTime.UtcNow.GetTimeMillis().ToString();
        }

        protected string _getDate()
        {
            return TimeUtils.GetGMTDate();
        }

        protected string _getContentMD5(TeaRequest request)
        {
            using(MD5 mi = MD5.Create())
            {
                byte[] buffer = mi.ComputeHash(request.Body);
                request.Body.Seek(0, System.IO.SeekOrigin.Begin);
                return Convert.ToBase64String(buffer);
            }
        }

        protected string _getHost()
        {
            return _domain;
        }

        protected string _getSignature(TeaRequest request)
        {
            string signedHeader = BaseUtils.GetSignedHeader(request);
            string url = BaseUtils.BuildUrl(request);
            string date = DictUtils.GetDicValue(request.Headers, "date").ToSafeString();
            string accept = DictUtils.GetDicValue(request.Headers, "accept").ToSafeString();
            string contentType = DictUtils.GetDicValue(request.Headers, "content-type").ToSafeString();
            string contentMd5 = DictUtils.GetDicValue(request.Headers, "content-md5").ToSafeString();
            string signStr = request.Method + "\n" + accept + "\n" + contentMd5 + "\n" + contentType + "\n" + date + "\n" + signedHeader + "\n" + url;

            byte[] signData;
            using(KeyedHashAlgorithm algorithm = CryptoConfig.CreateFromName("HMACSHA256") as KeyedHashAlgorithm)
            {
                algorithm.Key = Encoding.UTF8.GetBytes(_appSecret);
                signData = algorithm.ComputeHash(Encoding.UTF8.GetBytes(signStr.ToCharArray()));
            }
            string signedStr = Convert.ToBase64String(signData);
            return signedStr;
        }

        public bool _isFail(TeaResponse resp)
        {
            return resp.StatusCode < 200 || resp.StatusCode >= 300;
        }

        public string _default(string strValue, string strDefault)
        {
            if (string.IsNullOrWhiteSpace(strValue))
            {
                return strDefault;
            }
            return strValue;
        }

        public int? _defaultNumber(int? numValue, int? numDefault)
        {
            if (numValue > 0)
            {
                return numValue;
            }
            return numDefault;
        }

    }
}

File: php/src/BaseClient.php
<?php

namespace AlibabaCloud\ApiGateway\Util;

use AlibabaCloud\Tea\Request;

class BaseClient
{
    private static $filterKey = [
        'x-ca-signature',
        'x-ca-signature-headers',
        'accept',
        'content-md5',
        'content-type',
        'date',
        'host',
        'token',
        'user-agent',
    ];

    /**
     * @param string $secret
     *
     * @throws \Exception
     *
     * @return string
     */
    public static function getSignature(Request $request, $secret)
    {
        $signedHeader = self::_getSignedHeader($request);
        $url          = self::_buildUrl($request);
        $date         = isset($request->headers['date']) ? $request->headers['date'] : '';
        $accept       = isset($request->headers['accept']) ? $request->headers['accept'] : '';
        $contentType  = isset($request->headers['content-type']) ? $request->headers['content-type'] : '';
        $contentMd5   = isset($request->headers['content-md5']) ? $request->headers['content-md5'] : '';

        $signStr = implode("\n", [
            strtoupper($request->method),
            $accept,
            $contentMd5,
            $contentType,
            $date,
            $signedHeader,
            $url,
        ]);

        return base64_encode(hash_hmac('sha256', $signStr, $secret, true));
    }

    /**
     * @param array $query
     *
     * @throws \Exception
     *
     * @return array
     */
    public static function toQuery($query)
    {
        $res = [];
        foreach ($query as $key => $val) {
            if (null !== $val) {
                $res[$key] = $val;
            }
        }

        return $res;
    }

    /**
     * @param int $code
     *
     * @throws \Exception
     *
     * @return bool
     */
    public static function isFail($code)
    {
        return $code < 200 || $code >= 300;
    }

    /**
     * @param string $body
     *
     * @throws \Exception
     *
     * @return string
     */
    public static function getContentMD5($body)
    {
        return base64_encode(md5($body, true));
    }

    /**
     * @param Request $request
     *
     * @return string
     */
    private static function _getSignedHeader($request)
    {
        $headers = [];
        foreach ($request->headers as $key => $header) {
            $headers[strtolower($key)] = $header;
        }
        $resHeaders = [];
        ksort($headers);
        $keys = [];
        foreach ($headers as $key => $val) {
            if (!\in_array($key, self::$filterKey)) {
                array_push($keys, $key);
                array_push($resHeaders, $key . ':' . $val);
            }
        }
        $request->headers['x-ca-signature-headers'] = implode(',', $keys);

        return implode("\n", $resHeaders);
    }

    /**
     * @param Request $request
     *
     * @return string
     */
    private static function _buildUrl($request)
    {
        $url   = $request->pathname ? $request->pathname : '';
        $query = $request->query;
        ksort($query);
        if (\count($query) > 0) {
            if (false === strpos($url, '?')) {
                $url .= '?';
            }
            $tmp = [];
            foreach ($query as $key => $val) {
                array_push($tmp, $key . '=' . $val);
            }
            $url .= implode('&', $tmp);
        }

        return $url;
    }
}
